include/nbl/asset/IRayTracingPipeline.h

```diff
@@ -0,0 +1,173 @@
+#include <type_traits>
+
+namespace nbl::asset
+{
+
+class IRayTracingPipelineBase : public virtual core::IReferenceCounted
+{
+  public:
+    struct SShaderGroupsParams
+    {
+      struct SIndex
+      {
+        constexpr static inline uint32_t Unused = 0xffFFffFFu;
+        uint32_t index = Unused;
+      };
+
+      struct SHitGroup
+      {
+        uint32_t closestHit = SIndex::Unused;
+        uint32_t anyHit = SIndex::Unused;
+        uint32_t intersectionShader = SIndex::Unused;
+      };
+
+      SIndex raygen;
+      std::span<SIndex> misses;
+      std::span<SHitGroup> hits;
+      std::span<SIndex> callables;
+
+      inline uint32_t getShaderGroupCount() const
+      {
+        return 1 + hits.size() + misses.size() + callables.size();
+      }
+
+    };
+    using SGeneralShaderGroup = SShaderGroupsParams::SIndex;
+    using SHitShaderGroup = SShaderGroupsParams::SHitGroup;
+
+    struct SCachedCreationParams final
+    {
+      uint32_t maxRecursionDepth : 6 = 0;
+      uint32_t dynamicStackSize : 1 = false;
+    };
+};
+
+template<typename PipelineLayoutType, typename ShaderType>
+class IRayTracingPipeline : public IPipeline<PipelineLayoutType>, public IRayTracingPipelineBase
+{
+  public:
+
+    using SGeneralShaderGroupContainer = core::smart_refctd_dynamic_array<SGeneralShaderGroup>;
+    using SHitShaderGroupContainer = core::smart_refctd_dynamic_array<SHitShaderGroup>;
+
+    struct SCreationParams : IPipeline<PipelineLayoutType>::SCreationParams
+    {
+      protected:
+        using SpecInfo = ShaderType::SSpecInfo;
+        template<typename ExtraLambda>
+        inline bool impl_valid(ExtraLambda&& extra) const
+        {
+          if (!IPipeline<PipelineLayoutType>::SCreationParams::layout)
+            return false;
+
+          core::bitflag<ICPUShader::E_SHADER_STAGE> stagePresence = {};
+          for (const auto info : shaders)
+            if (info.shader)
+            {
+              if (!extra(info))
+                return false;
+              const auto stage = info.shader->getStage();
+              if ((stage & ~ICPUShader::E_SHADER_STAGE::ESS_ALL_RAY_TRACING)!=0)  
+                return false;
+              if (!std::has_single_bit<std::underlying_type_t<ICPUShader::E_SHADER_STAGE>>(stage))
+                return false;
+              stagePresence |= stage;
+            }
+
+          auto getShaderStage = [this](size_t index) -> ICPUShader::E_SHADER_STAGE
+            {
+              return shaders[index].shader->getStage();
+            };
+
+          if (shaderGroups.raygen.index >= shaders.size())
+            return false;
+          if (getShaderStage(shaderGroups.raygen.index) != ICPUShader::E_SHADER_STAGE::ESS_RAYGEN)
+            return false;
+
+          auto isValidShaderIndex = [this, getShaderStage](size_t index, ICPUShader::E_SHADER_STAGE expectedStage) -> bool
+            {
+              if (index == SShaderGroupsParams::SIndex::Unused)
+                return true;
```

Comment:
only if you don't have the NO_NULL flag for the particular stage (can take that as a boolean into the lambda)